gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineModuleJvmArgs.java

```diff
@@ -0,0 +1,204 @@
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.compile.JavaCompile;
+import org.gradle.api.tasks.testing.Test;
+import org.gradle.jvm.tasks.Jar;
+import org.gradle.process.CommandLineArgumentProvider;
+
+/**
+ * This plugin reuses the {@code Add-Exports} manifest entry to propagate and collect required exports
+ * from transitive dependencies, and applies them to compilation (for annotation processors) and
+ * execution (tests, javaExec, etc) for runtime dependencies.
+ */
+public final class BaselineModuleJvmArgs implements Plugin<Project> {
+
+    private static final String EXTENSION_NAME = "moduleJvmArgs";
+    private static final String ADD_EXPORTS_ATTRIBUTE = "Add-Exports";
+
+    private static final Splitter EXPORT_SPLITTER =
+            Splitter.on(' ').trimResults().omitEmptyStrings();
+
+    @Override
+    public void apply(Project project) {
+        project.getPluginManager().withPlugin("java", unused -> {
+            BaselineExportsExtension extension =
+                    project.getExtensions().create(EXTENSION_NAME, BaselineExportsExtension.class, project);
+
+            // javac isn't provided `--add-exports` args for the time being due to
+            // https://github.com/gradle/gradle/issues/18824
+            if (project.hasProperty("add.exports.to.javac")) {
+                project.getExtensions().getByType(SourceSetContainer.class).configureEach(sourceSet -> {
+                    JavaCompile javaCompile = project.getTasks()
+                            .named(sourceSet.getCompileJavaTaskName(), JavaCompile.class)
+                            .get();
+                    javaCompile
+                            .getOptions()
+                            .getCompilerArgumentProviders()
+                            // Use an anonymous class because tasks with lambda inputs cannot be cached
+                            .add(new CommandLineArgumentProvider() {
+                                @Override
+                                public Iterable<String> asArguments() {
+                                    // Annotation processors are executed at compile time
+                                    ImmutableList<String> arguments =
+                                            collectAnnotationProcessorExports(project, extension, sourceSet);
+                                    project.getLogger()
+                                            .debug(
+                                                    "BaselineModuleJvmArgs compiling {} on {} with exports: {}",
+                                                    javaCompile.getName(),
+                                                    project,
+                                                    arguments);
+                                    return arguments;
+                                }
+                            });
+                });
+            }
```

Comment:
block is ignored for now, hopefully something becomes of https://github.com/gradle/gradle/issues/18824